flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// remove all states logically
+		for (int i = 1; i <= 100; i += 2) {
+			totalStateSize--;
+			stateMap.remove(i, (long) i);
+			removeFromReferenceState(referenceStates, i, (long) i);
+		}
+		assertEquals(0, totalStateSize);
+		verifyState(referenceStates, stateMap);
+
+		verifySnapshotWithoutTransform(
+			expectedSnapshot1, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot1.release();
+	}
+
+	private void updateStateForConcurrentSnapshots(
+		Map<Long, Map<Integer, String>> referenceStates,
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap,
+		int updateCounter) {
+		ThreadLocalRandom random = ThreadLocalRandom.current();
+		// update and remove some states
+		Set<Long> removedNamespaceSet = new HashSet<>();
+		for (Map.Entry<Long, Map<Integer, String>> namespaceEntry : referenceStates.entrySet()) {
+			long namespace = namespaceEntry.getKey();
+			for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+				int key = keyEntry.getKey();
+				// namespace should be equal to key
+				assertEquals(namespace, key);
+				int op = random.nextInt(3);
+				switch (op) {
+					case 0:
+						stateMap.remove(key, namespace);
+						removedNamespaceSet.add(namespace);
+						break;
+					case 1:
+						String state = String.valueOf(random.nextLong());
+						stateMap.put(key, namespace, state);
+						keyEntry.setValue(state);
+						break;
+					default:
+						break;
+				}
+			}
+		}
+		for (long namespace : removedNamespaceSet) {
+			referenceStates.remove(namespace);
+		}
+
+		for (int i = 0; i < 100; i++) {
+			int key = updateCounter + i * 50;
+			long namespace = key;
+			String state = String.valueOf(key * namespace);
+			stateMap.put(key, namespace, state);
+			addToReferenceState(referenceStates, key, namespace, state);
+		}
+	}
+
+	private <K, N, S> void addToReferenceState(Map<N, Map<K, S>> referenceStates, K key, N namespace, S state) {
+		referenceStates.computeIfAbsent(namespace, (none) -> new HashMap<>()).put(key, state);
+	}
+
+	private <K, N, S> void removeFromReferenceState(Map<N, Map<K, S>> referenceStates, K key, N namespace) {
+		Map<K, S> keyMap = referenceStates.get(namespace);
+		if (keyMap == null) {
+			return;
+		}
+		keyMap.remove(key);
+		if (keyMap.isEmpty()) {
+			referenceStates.remove(namespace);
+		}
+	}
+
+	private <K, N, S> void verifyState(
+		Map<N, Map<K, S>> referenceStates,
+		CopyOnWriteSkipListStateMap<K, N, S> stateMap) {
+
+		// validates get(K, N)
+		for (Map.Entry<N, Map<K, S>> entry : referenceStates.entrySet()) {
+			N namespace = entry.getKey();
+			for (Map.Entry<K, S> keyEntry : entry.getValue().entrySet()) {
+				K key = keyEntry.getKey();
+				S state = keyEntry.getValue();
+				assertEquals(state, stateMap.get(key, namespace));
+				assertTrue(stateMap.containsKey(key, namespace));
```

Comment:
Wouldn't it be enough to ensure via a test that whenever the map contains a value, then `containsKey` will return `true`? That way we wouldn't have to test it for every access to the state.